subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/IncrementalResultStoringCompiler.java

```diff
@@ -152,4 +127,62 @@ private CompilerApiData getCompilerApiData(JavaCompileSpec spec, WorkResult resu
+        if (processors.isEmpty()) {
+            return new AnnotationProcessingData();
+        }
+        AnnotationProcessingData previousAnnotationProcessingData = null;
+        RecompilationSpec recompilationSpec = null;
+        if (result instanceof IncrementalCompilationResult) {
+            previousAnnotationProcessingData = ((IncrementalCompilationResult) result).getPreviousCompilationData().getAnnotationProcessingData();
+            recompilationSpec = ((IncrementalCompilationResult) result).getRecompilationSpec();
+            result = ((IncrementalCompilationResult) result).getCompilerResult();
+        }
+        Set<String> changedClasses = recompilationSpec == null ? Collections.emptySet() : recompilationSpec.getClassesToCompile();
+
+        if (result instanceof ApiCompilerResult) {
+            AnnotationProcessingResult processingResult = ((ApiCompilerResult) result).getAnnotationProcessingResult();
+            AnnotationProcessingData newAnnotationProcessingData = new AnnotationProcessingData(
+                processingResult.getGeneratedTypesWithIsolatedOrigin(),
+                processingResult.getAggregatedTypes(),
+                processingResult.getGeneratedAggregatingTypes(),
+                processingResult.getGeneratedResourcesWithIsolatedOrigin(),
+                processingResult.getGeneratedAggregatingResources(),
+                processingResult.getFullRebuildCause()
+            );
+            if (previousAnnotationProcessingData == null) {
+                return newAnnotationProcessingData;
+            }
+            return mergeAnnotationProcessingData(previousAnnotationProcessingData, newAnnotationProcessingData, changedClasses);
+        }
+        return new AnnotationProcessingData(
+            ImmutableMap.of(),
+            ImmutableSet.of(),
+            ImmutableSet.of(),
+            ImmutableMap.of(),
+            ImmutableSet.of(),
+            "the chosen compiler did not support incremental annotation processing"
+        );
+
+    }
+
+    private AnnotationProcessingData mergeAnnotationProcessingData(AnnotationProcessingData oldData, AnnotationProcessingData newData, Set<String> changedClasses) {
+        Map<String, Set<String>> generatedTypesByOrigin = new HashMap<>(oldData.getGeneratedTypesByOrigin());
+        changedClasses.forEach(generatedTypesByOrigin::remove);
+        generatedTypesByOrigin.putAll(newData.getGeneratedTypesByOrigin());
+        Map<String, Set<GeneratedResource>> generatedResourcesByOrigin = new HashMap<>(oldData.getGeneratedResourcesByOrigin());
+        changedClasses.forEach(generatedResourcesByOrigin::remove);
+        generatedResourcesByOrigin.putAll(newData.getGeneratedResourcesByOrigin());
+
+        return new AnnotationProcessingData(
+            generatedTypesByOrigin,
+            newData.getAggregatedTypes(),
+            newData.getGeneratedTypesDependingOnAllOthers(),
```

Comment:
What is the reason we don't need to merge also other data here?

Reply:
Aggregated types are always reprocessed, thus every compilation knows the full set of aggregated types.